libtenzir/builtins/operators/from_to_2.cpp

```diff
@@ -582,9 +645,702 @@ class to_plugin2 final : public virtual operator_factory_plugin {
+    if (file.IsFile() and matches(file.path(), glob_)) {
+      add_job(std::move(file));
+    }
+  }
+
+  void got_all_files() {
+    if (args_.watch) {
+      std::swap(previous_, current_);
+      current_.clear();
+      self_->run_delayed_weak(watch_pause, [this] {
+        query_files();
+      });
+    } else {
+      added_all_jobs_ = true;
+      check_termination();
+    }
+  }
+
+  void check_termination() {
+    if (added_all_jobs_ and jobs_.empty() and active_jobs_ == 0) {
+      for (auto& get : gets_) {
+        // If there are any unmatched gets, we know that there are no puts.
+        get.deliver(table_slice{});
+      }
+    }
+  }
+
+  void check_jobs() {
+    while (not jobs_.empty() and active_jobs_ < max_jobs) {
+      start_job(jobs_.front());
+      jobs_.pop_front();
+    }
+  }
+
+  void check_jobs_and_termination() {
+    check_jobs();
+    check_termination();
+  }
+
+  void add_job(arrow::fs::FileInfo file) {
+    auto inserted = current_.emplace(file).second;
+    TENZIR_ASSERT(inserted);
+    if (previous_.contains(file)) {
+      return;
+    }
+    jobs_.push_back(std::move(file));
+    check_jobs();
+  }
+
+  auto make_pipeline(std::string_view path) -> failure_or<pipeline> {
+    if (args_.pipe) {
+      return args_.pipe->inner;
+    }
+    auto path_dh = transforming_diagnostic_handler{
+      *dh_, [path](diagnostic d) {
+        // TODO: Should an error here be an error for `from_file`?
+        return std::move(d)
+          .modify()
+          .note(fmt::format("while reading `{}`", path))
+          .done();
+      }};
+    TRY(auto compression_and_format,
+        get_compression_and_format<true>(
+          located<std::string_view>{path, args_.url.source}, nullptr,
+          "https://docs.tenzir.com/tql2/operators/from_file", path_dh));
+    auto& format = compression_and_format.format.get();
+    auto compression = compression_and_format.compression;
+    auto provider = session_provider::make(path_dh);
+    auto ctx = provider.as_session();
+    auto pipe = pipeline{};
+    if (compression) {
+      TRY(auto decompress, make_operator(*compression, args_.url.source, ctx));
+      pipe.append(std::move(decompress));
+    }
+    TRY(auto read, make_operator(format, args_.url.source, ctx));
+    pipe.append(std::move(read));
+    return pipe;
+  }
+
+  void start_job(const arrow::fs::FileInfo& file) {
+    TENZIR_ASSERT(active_jobs_ < max_jobs);
+    active_jobs_ += 1;
+    auto pipe = make_pipeline(file.path());
+    if (pipe.is_error()) {
+      active_jobs_ -= 1;
+      return;
+    }
+    // We already checked the output type after parsing.
+    auto output_type = pipe->infer_type<chunk_ptr>();
+    TENZIR_ASSERT(output_type);
+    TENZIR_ASSERT(output_type->is<table_slice>());
+    add_actor_callback(
+      self_, fs_->OpenInputStreamAsync(file),
+      [this, pipe = std::move(*pipe), path = file.path()](
+        arrow::Result<std::shared_ptr<arrow::io::InputStream>> stream) mutable {
```

Comment:
This'd be much easier to follow if you moved the logic of this lambda it a member function and then just call that inside the lambda.